src/runtime/virtcontainers/qemu.go

```diff
@@ -1020,47 +1053,209 @@ func (q *qemu) togglePauseSandbox(pause bool) error {
 	if err != nil {
 		q.Logger().WithError(err).Error("Failed to connect to QEMU instance")
 		return err
 	}
 
-	err = qmp.ExecuteQMPCapabilities(q.qmpMonitorCh.ctx)
+	err = qmp.ExecuteQMPCapabilities(qmpChan.ctx)
 	if err != nil {
 		qmp.Shutdown()
 		q.Logger().WithError(err).Error(qmpCapErrMsg)
 		return err
 	}
-	q.qmpMonitorCh.qmp = qmp
-	q.qmpMonitorCh.disconn = disconnectCh
+	qmpChan.qmp = qmp
+	qmpChan.disconn = disconnectCh
+
+	return nil
+}
+
+func (q *qemu) loopQMPEvent(event chan govmmQemu.QMPEvent) {
+	for {
+		select {
+		case e, open := <-event:
+			if !open {
+				q.Logger().Infof("QMP event channel closed")
+				return
+			}
+			q.Logger().WithField("event", e).Debugf("got QMP event")
+			if e.Name == "GUEST_PANICKED" {
+				go q.handleGuestPanic()
+			}
+		}
+	}
+}
+
+func (q *qemu) handleGuestPanic() {
+	if err := q.dumpGuestMemory(q.config.GuestMemoryDumpPath); err != nil {
+		q.Logger().WithError(err).Error("failed to dump guest memory")
+	}
+
+	// TODO: how to notify the upper level sandbox to handle the error
+	// to do a fast fail(shutdown or others).
+	// tracked by https://github.com/kata-containers/kata-containers/issues/1026
+}
 
+// canDumpGuestMemory check if can do a guest memory dump operation.
+// for now it only ensure there must be double of VM size for free disk spaces
+func (q *qemu) canDumpGuestMemory(dumpSavePath string) error {
+	fs := unix.Statfs_t{}
+	if err := unix.Statfs(dumpSavePath, &fs); err != nil {
+		q.Logger().WithError(err).WithField("dumpSavePath", dumpSavePath).Error("failed to call Statfs")
+		return nil
+	}
+	availSpaceInBytes := fs.Bavail * uint64(fs.Bsize)
+	q.Logger().WithFields(
+		logrus.Fields{
+			"dumpSavePath":      dumpSavePath,
+			"availSpaceInBytes": availSpaceInBytes,
+		}).Info("get avail space")
+
+	// get guest memory size
+	guestMemorySizeInBytes := (uint64(q.config.MemorySize) + uint64(q.state.HotpluggedMemory)) << utils.MibToBytesShift
+	q.Logger().WithField("guestMemorySizeInBytes", guestMemorySizeInBytes).Info("get guest memory size")
+
+	// default we want ensure there are at least double of VM memory size free spaces available,
+	// this may complete one dump operation for one sandbox
+	exceptMemorySize := guestMemorySizeInBytes * 2
+	if availSpaceInBytes >= exceptMemorySize {
+		return nil
+	} else {
+		return fmt.Errorf("there are not enough free space to store memory dump file. Except %d bytes, but only %d bytes available", exceptMemorySize, availSpaceInBytes)
+	}
+}
+
+// dumpSandboxMetaInfo save meta information for debug purpose, includes:
+// hypervisor verison, sandbox/container state, hypervisor config
+func (q *qemu) dumpSandboxMetaInfo(dumpSavePath string) {
+	dumpStatePath := filepath.Join(dumpSavePath, "state")
+
+	// copy state from /run/vc/sbs to memory dump directory
+	statePath := filepath.Join(q.store.RunStoragePath(), q.id)
+	command := []string{"cp", "-r", statePath, dumpStatePath}
```

Comment:
I'm not totally comfortable about this for a number of reasons:

- Shouldn't we request the current state be persisted to disk before copying the on-disk representation (which may be out of date)?
- I'd rather this method calls a public `store` package function to get the metadata rather than using the on-disk data.
- Do we need all the state I wonder?
- If we have to use `cp`, it needs to be `/bin/cp` and probably `/bin/cp -a`.

/cc @WeiZhang555, @bergwolf.

Reply:
Hi @jodh-intel 
1. At the time  kernel panic, the state should be  the latest because when sandbox state changes, the `s.storeSandbox()`will be called.
2. `store` package provide function `FromDisk(sid string) (SandboxState, map[string]ContainerState, error)`,  this may need us to marshal these to write to filesystem.
3. Select what is needed needs more code, save all state is the simplest, and the state files dose not need much disk spaces.
4. Done.